### PR TITLE
Feature/storage use local read

### PR DESCRIPTION
**Pull request by achocron**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

This PR contains the logic to read from local storage if appropriate.

### CODE REVIEW

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 1]
```diff
@@ -66,6 +71,8 @@ - (instancetype)init {
   if ((self = [super init])) {
```
Comment (MehrdadMzfr) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: pass `MS_Reachability` as and argument so that you can mock it in unit test

Comment (YiMysty) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: I think we already did that? :)

Comment (YiMysty) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: in the test case ```  // Simulate being offline.
  MS_Reachability *reachabilityMock = OCMPartialMock([MS_Reachability reachabilityForInternetConnection]);
  OCMStub([reachabilityMock currentReachabilityStatus]).andReturn(NotReachable);
  self.sut.reachability = reachabilityMock;```

**Comment by clpolet** [comment]
One test unrelated test crashed.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 2]
```diff
@@ -225,34 +232,109 @@ - (NSError *)generateDisabledError:(NSString *)operation documentId:(NSString *)
...
+                                               NSLocalizedDescriptionKey : @"Cannot read from local storage because there is no "
+                                                                           @"account ID cached and failed to retrieve token."
+                                             }];
+                                   MSDocumentWrapper *documentWrapper = [[MSDocumentWrapper alloc] initWithError:authenticationError
+                                                                                                      documentId:documentId];
+                                   completionHandler(documentWrapper);
+                                   return;
+                                 }
+
+                                 // Run the operation in a dispatch queue to avoid blocking and concurrent database accesses.
```
Comment (clpolet) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Maybe rephrase?

File: AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m [hunk 3]
```diff
@@ -58,12 +58,14 @@ @implementation MSDataStoreTests
...
+static NSString *const kMSAccountId = @"ceb61029-d032-4e7a-be03-2614cfe2a564";
 static NSString *const kMSDbNameTest = @"dbName";
 static NSString *const kMSDbCollectionNameTest = @"dbCollectionName";
 static NSString *const kMSStatusTest = @"status";
-static NSString *const kMSExpiresOnTest = @"20191212";
+static NSString *const kMSExpiresOnTest = @"2999-09-19T11:11:11.111Z";
+;
 static NSString *const kMSDocumentIdTest = @"documentId";
 
 - (void)setUp {
```
Comment (clpolet) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m]: Since Datastore is now using a dispatch queue, we'd like to make sure the queue doesn't execute any other code while the test is done. There is an example there: https://github.com/Microsoft/AppCenter-SDK-Apple/blob/a3dee08381714c300d501f8de61531cf58024d05/AppCenter/AppCenterTests/MSChannelGroupDefaultTests.m#L54

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 1 above)
Comment (MehrdadMzfr) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: it works as you implemented, yes it should not be a public setter. I was suggesting to pass is as constructor argument

File: AppCenter/AppCenterTests/Util/MSDispatchTestUtil.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MSDispatchTestUtil : NSObject
+
+/**
+ * Suspends the dispatch queue and waits for current operations to complete.
+ *
+ * @param dispatchQueue The dispatch queue to suspend.
+ */
++ (void) awaitAndSuspendDispatchQueue:(dispatch_queue_t)dispatchQueue;
```
Comment (MehrdadMzfr) [inline AppCenter/AppCenterTests/Util/MSDispatchTestUtil.h]: NIT: space between ) await..

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 5]
```diff
@@ -225,34 +232,109 @@ - (NSError *)generateDisabledError:(NSString *)operation documentId:(NSString *)
...
-                                    [cosmosDbError localizedDescription]);
-                         completionHandler([[MSDocumentWrapper alloc] initWithError:cosmosDbError documentId:documentId]);
-                         return;
-                       }
+    [self readFromLocalStorageWithPartition:partition
+                                 documentId:documentId
+                               documentType:documentType
+                                readOptions:readOptions
+                          completionHandler:^(MSDocumentWrapper *_Nonnull document) {
+                            if ([self.reachability currentReachabilityStatus] == NotReachable) {
```
Comment (MehrdadMzfr) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: combine 2 ifs

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 6]
```diff
@@ -225,34 +232,109 @@ - (NSError *)generateDisabledError:(NSString *)operation documentId:(NSString *)
...
+                                               userInfo:@{
+                                                 NSLocalizedDescriptionKey : @"The document was marked for deletion and could not be read."
+                                               }];
+                                     documentWrapper = [[MSDocumentWrapper alloc] initWithError:notFoundError documentId:documentId];
+                                     documentWrapper.pendingOperation = kMSPendingOperationDelete;
+                                   }
+                                   completionHandler(documentWrapper);
+                                 });
+                               }];
+  return;
```
Comment (MehrdadMzfr) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: not needed, remove

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 7]
```diff
@@ -225,34 +232,109 @@ - (NSError *)generateDisabledError:(NSString *)operation documentId:(NSString *)
...
+                                  [cosmosDbError localizedDescription]);
+                       completionHandler([[MSDocumentWrapper alloc] initWithError:cosmosDbError documentId:documentId]);
                        return;
-                     }];
-  }
+                     }
+
+                     // Deserialize.
+                     completionHandler([MSDocumentUtils documentWrapperFromData:data documentType:documentType]);
+                     return;
```
Comment (MehrdadMzfr) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: not needed

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m [hunk 8]
```diff
@@ -102,7 +102,7 @@ - (BOOL)deleteUserStorageWithAccountId:(NSString *)accountId {
 - (MSDocumentWrapper *)readWithPartition:(NSString *)partition
                               documentId:(NSString *)documentId
                             documentType:(Class)documentType
-                             readOptions:(MSReadOptions *)__unused readOptions {
+                             readOptions:(nullable MSReadOptions *)__unused readOptions {
   NSString *selectionQuery =
       [NSString stringWithFormat:@"SELECT * FROM \"%@\" WHERE \"%@\" = \"%@\" AND \"%@\" = \"%@\"", kMSAppDocumentTableName,
```
Comment (MehrdadMzfr) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: the user table name is in this format, you should use a util method to get the table name
e.g. 
```
+ (NSString *)userTableNameWithAccountId:(NSString *)accountId {
  return [NSString stringWithFormat:kMSUserDocumentTableNameFormat, accountId];
}
```

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m [hunk 9]
```diff
@@ -147,22 +149,24 @@ + (void)saveToken:(MSTokenResult *)tokenResult {
...
     if (tokenString) {
       MSTokenResult *tokenResult = [[MSTokenResult alloc] initWithString:tokenString];
       NSDate *currentUTCDate = [NSDate date];
       NSDate *tokenExpireDate = [MSUtility dateFromISO8601:tokenResult.expiresOn];
       if ([currentUTCDate laterDate:tokenExpireDate] == currentUTCDate) {
-        [MSTokenExchange removeCachedToken:partitionName];
-        MSLogWarning([MSDataStore logTag], @"The token in the cache has expired for the partition : %@.", partitionName);
+        MSLogWarning([MSDataStore logTag], @"The token in the cache has expired for the partition : %@.", partition);
+        if (expiredTokenIncluded) {
+          return tokenResult;
```
Comment (Jamminroot) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m]: I think this line is uncovered

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m [hunk 10]
```diff
@@ -6,6 +6,7 @@
 #import "MSAuthTokenContext.h"
 #import "MSConstants+Internal.h"
 #import "MSDataStorageConstants.h"
+#import "MSDataStore.h"
```
Comment (Jamminroot) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m]: I think this can be removed, since `MSDataStoreInternal.h` already included

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m (hunk 9 above)
Comment (achocron) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m]: I think it should be - but I think that this is unrelated to this PR

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 1 above)
Comment (achocron) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: It's more like an internal setter because it isn't in the public header file

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m (hunk 8 above)
Comment (achocron) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: Will fix this in a different PR
